CI/DOC Use sphinx-gallery 0.3.1

lesteve opened the pull request:
This PR target the `0.21.X` branch. I cherry-picked the commit from #14507 (merged 4 weeks ago in master).

My goal behind this is to add binder links to the stable doc (very similar to #11221 which added binder links to the dev doc). The work done in master needs sphinx-gallery 0.3.1.

I am not expecting any issues, but once the doc is built in this PR I will double-check whether I don't see any obvious problems in the generated doc.

Also what is the preferred way of changing things in a relase branch:
* should this PR be merged (similar workflow as master)?
* should someone push the commit directly in the `0.21.X` branch (more obvious that it is a cherry-pick of a commit in master)? This seems to be the approach for bug-fixes in `0.21.X` as far as I can see.

lesteve committed:
CI/DOC Use sphinx-gallery 0.3.1 (#14507)

[doc build]

jnothman commented:
I'm fine with merging to the branch... Bug fix releases are cherry-picked
into a branch and then merged without squash in order to avoid having the
docs inconsistent with the released code, while maintaining individuated
commits.

lesteve committed:
Switch to sphinx 1.6.3.

sphinx 1.6.2 is only available in free channel (which was removed).

[doc build]

lesteve commented:
I compared the [stable doc](https://scikit-learn.org/stable/auto_examples/) and the [doc generated from this PR](https://71909-843222-gh.circle-artifacts.com/0/doc/auto_examples/index.html) and I did not spot any difference.

It seems like I will have to fix a few things related to conda removing the `free` channel by default though to get the CIs green again in 0.21.X ...

lesteve commented:
Closing this one, waiting for the next scikit-learn relase for the Binder links to appear in the stable doc seems a better idea.
